### PR TITLE
Harden MCP filesystem policy and command runtime defaults

### PR DESCRIPTION
**Pull request by rmusser01**
<h3>PR Summary by Qodo</h3>

Harden MCP filesystem policy and command runtime defaults
<code>✨ Enhancement</code> <code>🐞 Bug fix</code> <code>🧪 Tests</code> <code>📝 Documentation</code> <code>🕐 20-40 Minutes</code>

Grey Divider

<h3>Walkthroughs</h3>

<details open>
<summary>User Description</summary>

<br/>

## Summary
- Forward module-derived path scope candidates through the tldw MCP policy adapter.
- Add regression coverage for action-aware filesystem path grants, deny overrides, and mixed patch candidates.
- Prefer structured `fs.read` and `fs.write` create routing in the virtual command runtime while preserving legacy fallbacks and visible-tool filtering.
- Close TASK-2331 with verification notes and document deferred `write-replace` scope.

## Test Plan
- [x] `source /Users/macbook-dev/Documents/GitHub/tldw_server2/.venv/bin/activate && python -m pytest tldw_Server_API/app/core/MCP_unified/tests/test_protocol_path_scope_candidates.py tldw_Server_API/tests/MCP_unified/test_mcp_protocol_path_scope.py tldw_Server_API/app/core/MCP_unified/tests/test_command_runtime_registry.py tldw_Server_API/app/core/MCP_unified/tests/test_run_command_module.py -q` - 57 passed, 6 warnings
- [x] `source /Users/macbook-dev/Documents/GitHub/tldw_server2/.venv/bin/activate && python -m bandit -r tldw_Server_API/app/core/MCP_unified/adapters/tldw_policy.py tldw_Server_API/app/core/MCP_unified/command_runtime/registry.py tldw_Server_API/app/core/MCP_unified/command_runtime/adapters.py tldw_Server_API/app/core/MCP_unified/modules/implementations/run_command_module.py -f json -o /tmp/bandit_mcp_fs_policy_hardening_final.json` - 0 findings
- [x] `git diff --check`

## Change summary
Maintainer note: this PR was materially AI-authored. Replace this section with a human-written Change summary before merge per `Docs/superpowers/AI_GENERATED_PR_CHANGE_SUMMARY_POLICY_2026_04_17.md`.

</details>

<details open>
<summary>AI Description</summary>

<br/>

<pre>
• Fix <b><i>TldwPathScopeEnforcer.evaluate_tool_call</i></b> to accept and forward <b><i>path_scope_candidates</i></b> to
  the underlying enforcement service, closing a seam that caused <b><i>fs.patch</i></b> to bypass module-derived
  path/action context.
• Update the virtual CLI command registry so <b><i>cat</i></b> prefers <b><i>fs.read</i></b> over legacy <b><i>fs.read_text</i></b>
  (with fallback), and add a new <b><i>write-create</i></b> command backed by <b><i>fs.write</i></b> with <b><i>mode: &quot;create&quot;</i></b>.
• Filter <b><i>visible_commands</i></b> to return only currently-allowed backend tools per descriptor, enabling
  adapters to reliably select the preferred structured tool.
• Add <b><i>fs.write</i></b> to the write-classification set in <b><i>RunCommandModule</i></b> so <b><i>write-create</i></b> is
  correctly flagged for approval/audit.
• Add regression tests covering action-aware path grants (read/edit/write isolation), deny-override
  for nested prefixes, and patch-bundle fail-closed behavior.
• Add design spec, implementation plan, and backlog task closure docs for TASK-2331.
</pre>
</details>

<details>
<summary>Diagram</summary>

<br/>

```mermaid
graph TD
    A["run / bash / shell\n(RunCommandModule)"] --> B["CommandRegistry\n(registry.py)"]
    B --> C["visible_commands\n(filtered descriptors)"]
    C --> D["PhaseOneCommandAdapters\n(adapters.py)"]
    D --> E["fs.read / fs.read_text\n(cat routing)"]
    D --> F["fs.write mode=create\n(write-create)"]
    D --> G["fs.write_text\n(legacy write)"]
    H["MCP Protocol\n(fs.patch / fs.edit)"] --> I["TldwPathScopeEnforcer\n(tldw_policy.py)"]
    I --> J[("McpHubPathEnforcementService")]
    I -.->|"path_scope_candidates\n(NEW)"|J

    subgraph Legend
      direction LR
      _svc["Module / Service"] ~~~ _db[("Enforcement Service")] ~~~ _tool["Tool Endpoint"]
    end
```

</details>

<details>
<summary>High-Level Assessment</summary>

<br/>

The PR&#x27;s approach is well-suited to the problem. The seam fix (forwarding `path_scope_candidates`) is the minimal correct change — the enforcement service already had the capability, the adapter was simply dropping the context. The registry filtering approach (returning cloned descriptors with only visible backend tools) is clean and avoids duplicating policy logic in adapters. Deferring `write-replace` until preimage authorization syntax is available is the right call. No meaningfully better alternatives were identified.

</details>

Grey Divider

<h3>File Changes</h3>

<details>
<summary><strong>Enhancement</strong> (2)</summary>

<blockquote>
<details>
<summary><strong>registry.py</strong> <code>Add fs.read preference and write-create command to registry</code> <code>+13/-4</code></summary>

<br/>

>Add fs.read preference and write-create command to registry
>
><pre>
>• Updates &#x27;cat&#x27; backend tools to &#x27;(&quot;fs.read&quot;, &quot;fs.read_text&quot;)&#x27; so the structured tool is preferred. Adds a new &#x27;write-create&#x27; descriptor backed by &#x27;fs.write&#x27;. Refactors &#x27;visible_commands&#x27; to return filtered descriptors containing only currently-allowed backend tools, enabling adapters to select the best available tool without separate policy inspection. Also adds &#x27;mcp.catalogs.list&#x27; to the &#x27;mcp&#x27; command&#x27;s backend tools.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-75aade3d45d0aa309ae1148c3bbf17a047bdb4f1c6a9e2b24e49acb7f4fa8902'>tldw_Server_API/app/core/MCP_unified/command_runtime/registry.py</a>
<hr/>

</details>
</blockquote>

<blockquote>
<details>
<summary><strong>adapters.py</strong> <code>Route cat to fs.read and add write-create command adapter</code> <code>+61/-2</code></summary>

<br/>

>Route cat to fs.read and add write-create command adapter
>
><pre>
>• Updates &#x27;_governed_plan&#x27; for &#x27;cat&#x27; to use &#x27;_visible_backend_tool&#x27; helper to prefer &#x27;fs.read&#x27; over &#x27;fs.read_text&#x27;. Adds &#x27;write-create&#x27; command routing to &#x27;fs.write&#x27; with &#x27;mode: &quot;create&quot;&#x27;. Adds &#x27;_visible_backend_tool&#x27;, &#x27;_command_has_visible_backend&#x27;, and &#x27;_unavailable_backend_error&#x27; helpers. Applies backend-availability guards to &#x27;knowledge&#x27;, &#x27;media&#x27;, and &#x27;mcp&#x27; sub-commands. Enhances &#x27;_render_cat&#x27; to append a truncation marker when &#x27;fs.read&#x27; returns a truncated response.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-5a9dcd758151412dd96bdf2c1b640667e4416425511e0a0a7dd2de4a700c6796'>tldw_Server_API/app/core/MCP_unified/command_runtime/adapters.py</a>
<hr/>

</details>
</blockquote>

</details>

<details>
<summary><strong>Bug fix</strong> (2)</summary>

<blockquote>
<details>
<summary><strong>tldw_policy.py</strong> <code>Forward path_scope_candidates through TldwPathScopeEnforcer</code> <code>+2/-0</code></summary>

<br/>

>Forward path_scope_candidates through TldwPathScopeEnforcer
>
><pre>
>• Adds an optional &#x27;path_scope_candidates&#x27; parameter to &#x27;TldwPathScopeEnforcer.evaluate_tool_call&#x27; and forwards it to the underlying &#x27;McpHubPathEnforcementService&#x27;. This closes a seam where module-derived path/action candidates for &#x27;fs.patch&#x27; were silently dropped before reaching the real enforcement service.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-12ec74522daab857ee8a6f7a44ace3113f59661346bff2432ec28762600bea35'>tldw_Server_API/app/core/MCP_unified/adapters/tldw_policy.py</a>
<hr/>

</details>
</blockquote>

<blockquote>
<details>
<summary><strong>run_command_module.py</strong> <code>Add fs.write to write-classification set for approval/audit</code> <code>+1/-1</code></summary>

<br/>

>Add fs.write to write-classification set for approval/audit
>
><pre>
>• Adds &#x27;fs.write&#x27; to &#x27;_RUN_WRITE_BACKEND_TOOLS&#x27; so that &#x27;write-create&#x27; commands backed by the structured write primitive are correctly classified as write-capable for approval gating and audit purposes.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-54be7292069fca961e1d1c62ab0994567bb5a60516b83c4e8459688d3c0db933'>tldw_Server_API/app/core/MCP_unified/modules/implementations/run_command_module.py</a>
<hr/>

</details>
</blockquote>

</details>

<details>
<summary><strong>Tests</strong> (4)</summary>

<blockquote>
<details>
<summary><strong>test_command_runtime_registry.py</strong> <code>Update registry tests for new backend mappings and visibility filtering</code> <code>+32/-2</code></summary>

<br/>

>Update registry tests for new backend mappings and visibility filtering
>
><pre>
>• Updates &#x27;test_registry_exposes_phase_one_mappings&#x27; to assert the new &#x27;cat&#x27;, &#x27;write-create&#x27;, and &#x27;mcp&#x27; backend tool tuples. Adds &#x27;test_registry_filters_visible_backend_tools_for_multi_backend_commands&#x27; and &#x27;test_registry_filters_visible_backend_tools_for_mcp_catalogs&#x27; to verify that &#x27;visible_commands&#x27; returns filtered descriptors per allowed tool set.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-f7f656fb9759a11ce43ac8b83108b0823585d269c3ad4517f8336f7f4ffd0697'>tldw_Server_API/app/core/MCP_unified/tests/test_command_runtime_registry.py</a>
<hr/>

</details>
</blockquote>

<blockquote>
<details>
<summary><strong>test_protocol_path_scope_candidates.py</strong> <code>Add integration test for TldwPathScopeEnforcer candidate forwarding</code> <code>+40/-0</code></summary>

<br/>

>Add integration test for TldwPathScopeEnforcer candidate forwarding
>
><pre>
>• Adds &#x27;test_tldw_path_scope_enforcer_forwards_module_candidates&#x27; which monkeypatches the path enforcement service and verifies that &#x27;PathScopeCandidate&#x27; values passed to &#x27;TldwPathScopeEnforcer.evaluate_tool_call&#x27; are forwarded to the underlying service.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-5da9a9f1d8a2854ed0e842098a55d3c7c42f693f43538744e257a3570d964fff'>tldw_Server_API/app/core/MCP_unified/tests/test_protocol_path_scope_candidates.py</a>
<hr/>

</details>
</blockquote>

<blockquote>
<details>
<summary><strong>test_run_command_module.py</strong> <code>Expand run module tests for structured read/write and truncation</code> <code>+170/-5</code></summary>

<br/>

>Expand run module tests for structured read/write and truncation
>
><pre>
>• Updates the protocol stub to expose &#x27;fs.read&#x27; and &#x27;fs.write&#x27; and adds corresponding execution handlers. Adds tests for: legacy &#x27;fs.read_text&#x27; fallback when &#x27;fs.read&#x27; is hidden; truncation metadata rendering from &#x27;fs.read&#x27;; &#x27;write-create&#x27; routing to &#x27;fs.write&#x27; with &#x27;mode: create&#x27;; help visibility for &#x27;write-create&#x27; vs legacy &#x27;write&#x27;; &#x27;mcp catalogs&#x27; unavailability when &#x27;mcp.catalogs.list&#x27; is hidden; and write-classification detection for &#x27;write-create&#x27;.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-0d553336abee6a3227f3a9e03a55bd67e2a611f50f04a149fd5f1a37e3c39497'>tldw_Server_API/app/core/MCP_unified/tests/test_run_command_module.py</a>
<hr/>

</details>
</blockquote>

<blockquote>
<details>
<summary><strong>test_mcp_protocol_path_scope.py</strong> <code>Add action-aware path grant regression tests at service boundary</code> <code>+211/-0</code></summary>

<br/>

>Add action-aware path grant regression tests at service boundary
>
><pre>
>• Adds three new async tests exercising &#x27;McpHubPathEnforcementService&#x27; directly: &#x27;test_path_grants_keep_read_edit_and_write_distinct&#x27; verifies read/edit/write grants are strictly isolated; &#x27;test_path_grants_deny_override_wins_for_nested_prefix&#x27; verifies deny grants on nested prefixes override broader allow grants; &#x27;test_path_grants_patch_bundle_fails_closed_when_create_needs_write&#x27; verifies that a patch bundle with a create candidate fails closed when only an edit grant is present.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-48b2a59b17abab35c89f728bd4f635930d08bc21fd638ae2e2972038d39bfc84'>tldw_Server_API/tests/MCP_unified/test_mcp_protocol_path_scope.py</a>
<hr/>

</details>
</blockquote>

</details>

<details>
<summary><strong>Documentation</strong> (3)</summary>

<blockquote>
<details>
<summary><strong>task-2331 - Harden-MCP-filesystem-path-policy-and-command-runtime-defaults.md</strong> <code>Close TASK-2331 with verification notes and deferred scope</code> <code>+88/-0</code></summary>

<br/>

>Close TASK-2331 with verification notes and deferred scope
>
><pre>
>• New backlog task file documenting TASK-2331 as Done, recording implementation commit SHAs, test/Bandit verification results, and explicitly noting that &#x27;write-replace&#x27; is deferred because preimage syntax is not implemented in this slice.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-aeed0cb23b6f2269bb50f30991f90f1d5d6474d18da500657f73e78f1df7bfe3'>backlog/tasks/task-2331 - Harden-MCP-filesystem-path-policy-and-command-runtime-defaults.md</a>
<hr/>

</details>
</blockquote>

<blockquote>
<details>
<summary><strong>2026-06-09-mcp-filesystem-policy-hardening-design.md</strong> <code>Add MCP filesystem policy hardening design spec</code> <code>+186/-0</code></summary>

<br/>

>Add MCP filesystem policy hardening design spec
>
><pre>
>• New design document covering the goals, non-goals, policy regression coverage requirements, candidate forwarding boundary, virtual CLI routing strategy, and open follow-ups for this hardening slice.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-085dbecdb26294943d452fe2713510c69878f7d3ff92ce5d5d69f4aad15c97ea'>Docs/superpowers/specs/2026-06-09-mcp-filesystem-policy-hardening-design.md</a>
<hr/>

</details>
</blockquote>

<blockquote>
<details>
<summary><strong>2026-06-09-mcp-filesystem-policy-hardening-implementation-plan.md</strong> <code>Add step-by-step implementation plan for MCP filesystem hardening</code> <code>+542/-0</code></summary>

<br/>

>Add step-by-step implementation plan for MCP filesystem hardening
>
><pre>
>• New implementation plan document with four tasks (adapter seam, regression tests, virtual CLI routing, validation/closeout), each broken into numbered steps with code snippets, expected test outcomes, and commit commands.
></pre>
>
><a href='https://github.com/rmusser01/tldw_server/pull/2327/files#diff-1a5f9b3e48c984782099cba8b61e01076d165f1bc96896bb500f68a37d8021b8'>Docs/superpowers/plans/2026-06-09-mcp-filesystem-policy-hardening-implementation-plan.md</a>
<hr/>

</details>
</blockquote>

</details>

Grey Divider

<a href="https://www.qodo.ai">Qodo Logo</a>

### CODE REVIEW

File: tldw_Server_API/app/core/MCP_unified/command_runtime/adapters.py [hunk 1]
```diff
@@ -307,13 +318,32 @@ def _governed_plan(self, argv: list[str]) -> _GovernedCallPlan | _UsageError:
...
+
     def _knowledge_plan(self, argv: list[str]) -> _GovernedCallPlan | _UsageError:
         if len(argv) < 2:
             return _UsageError("usage: knowledge <search|get> ...")
         sub = argv[1]
         if sub == "search":
             if len(argv) < 3:
                 return _UsageError("usage: knowledge search <query>")
+            if not self._command_has_visible_backend("knowledge", "knowledge.search"):
+                return self._unavailable_backend_error("knowledge.search")
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/MCP_unified/command_runtime/adapters.py:346]: Verified against current code in `3addabaa72`: hidden `knowledge` commands already return `Unknown command` with exit 127 before `_knowledge_plan()` because `execute()` and `preflight_chain()` require the command descriptor in `visible_commands`. Added regression coverage in `test_run_returns_unknown_for_hidden_multi_backend_commands`; no implementation change was needed.

File: tldw_Server_API/app/core/MCP_unified/command_runtime/adapters.py [hunk 2]
```diff
@@ -356,18 +392,24 @@ def _mcp_plan(self, argv: list[str]) -> _GovernedCallPlan | _UsageError:
             return _UsageError("usage: mcp <tools|modules|catalogs>")
         sub = argv[1]
         if sub == "tools":
+            if not self._command_has_visible_backend("mcp", "mcp.tools.list"):
+                return self._unavailable_backend_error("mcp.tools.list")
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/MCP_unified/command_runtime/adapters.py:396]: Verified against current code in `3addabaa72`: hidden `mcp` commands already return `Unknown command` with exit 127 before `_mcp_plan()` because `execute()` and `preflight_chain()` require the command descriptor in `visible_commands`. Added regression coverage in `test_run_returns_unknown_for_hidden_multi_backend_commands`; hidden subtools still return the existing unavailable-backend error.

File: tldw_Server_API/tests/MCP_unified/test_mcp_protocol_path_scope.py [hunk 3]
```diff
@@ -779,6 +811,185 @@ def test_path_scope_key_includes_active_workspace_id() -> None:
...
+        assert decision["reason_code"] == "path_action_not_granted"  # nosec B101
+        assert decision["redacted"] is True  # nosec B101
+
+    read_policy_grants = [{"prefix": "docs", "actions": ["read"]}]
+    read_allowed = await _evaluate(read_policy_grants, "fs.read", "read")
+    read_edit_denied = await _evaluate(read_policy_grants, "fs.edit", "edit")
+    read_write_denied = await _evaluate(read_policy_grants, "fs.write", "write")
+
+    assert read_allowed["within_scope"] is True  # nosec B101
+    assert read_allowed["path_decisions"][0]["requested_action"] == "read"  # nosec B101
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/MCP_unified/test_mcp_protocol_path_scope.py]: Fixed in `3addabaa72`: allowed-path assertions now read decisions from `scope_payload["path_decisions"]`, matching the service payload contract used elsewhere in the test file.
